packages/core/src/hooks/sign.ts

```diff
@@ -0,0 +1,118 @@
+import { TypedData } from 'starknet/utils/typedData'
+import type { Signature } from 'starknet'
+import { useCallback, useReducer } from 'react'
+import { Connector } from '..'
+
+interface State {
+  data?: string[]
+  error?: string
+  isLoading: boolean
+  isSuccess: boolean
+}
+
+interface StartSigning {
+  type: 'start_signing'
+}
+
+interface SetSignature {
+  type: 'set_signature'
+  data: Signature
+}
+
+interface SetSigningError {
+  type: 'set_error'
+  error: string
+}
+
+interface SetSuccess {
+  type: 'set_success'
+  isSuccess: boolean
+}
+
+interface Reset {
+  type: 'reset'
+}
+
+type Action = StartSigning | SetSignature | SetSigningError | SetSuccess | Reset
+
+function starknetSignReducer(state: State, action: Action): State {
+  if (action.type === 'start_signing') {
+    return {
+      ...state,
+      isLoading: true,
+    }
+  } else if (action.type === 'set_signature') {
+    return {
+      ...state,
+      data: action.data,
+      isLoading: false,
+      isSuccess: true,
+    }
+  } else if (action.type === 'set_error') {
+    return {
+      ...state,
+      error: action.error,
+      isLoading: false,
+      isSuccess: false,
+    }
+  } else if (action.type === 'reset') {
+    return {
+      ...state,
+      data: undefined,
+      error: undefined,
+      isLoading: false,
+      isSuccess: false,
+    }
+  }
+  return state
+}
+
+export interface UseSignTypedData {
+  data?: string[]
+  error?: string
+  isError: boolean
+  isIdle: boolean
+  isLoading: boolean
+  isSuccess: boolean
+  signTypedData: () => Promise<Signature | undefined>
+  reset: () => void
+}
+
+export function useSignTypedData(connector: Connector, typedData: TypedData): UseSignTypedData {
+  const [state, dispatch] = useReducer(starknetSignReducer, {
+    isLoading: false,
+    isSuccess: false,
+  })
```

Comment:
```suggestion
export function useSignTypedData(typedData: TypedData): UseSignTypedData {
  const { account } = useStarknet()
  
  if (!account) {
     // error
  }
  
  const [state, dispatch] = useReducer(starknetSignReducer, {
    isLoading: false,
    isSuccess: false,
  })
```

Reply:
I think this suggestion breaks the rule of hooks.

Reply:
ah yeah the error should be returned after the other hooks

Reply:
I added this but because `useStarknet` does not return the connected `Connector` or any `AccountInterface`, I did the following. 

```typescript
  const signTypedData = useCallback(async () => {
    dispatch({ type: 'reset' })
    dispatch({ type: 'start_signing' })
    try {
      let accountInterface: AccountInterface | null = null
      for (const connector of connectors) {
        const account = await connector.account()
        if (account.address === accountAddress) {
          accountInterface = account
          break
        }
      }
      if (!accountInterface) {
        throw new Error(`No connector for address ${accountAddress}`)
      }
      const response = await accountInterface.signMessage(typedData)

      ...
```

This works but it would be ideal if I could access [this connector](https://github.com/apibara/starknet-react/blob/5d9503bea4c8c706684225057239e3e235f39272/packages/core/src/providers/starknet/manager.ts#L11) that's stored and used for `disconnect()`, but as far as I can tell I cannot. We can add it to `StarknetState` but that probably needs a follow up issue.

Reply:
hm yeah it feels like we should have a way to access the currently selected connector/account interface

Reply:
I can add accessing the current connector in #127.